tests/0102-static_group_rebalance.c

```diff
@@ -0,0 +1,217 @@
+        if (timeout_s > 0) {
+                TEST_SAY("%s: simulate processing by sleeping for %ds\n",
+                         rd_kafka_name(cons->rk), timeout_s);
+                rd_sleep(timeout_s);
+        }
+}
+
+/* Serve op queue until we have an assignment */
+static void await_assignment (_consumer_t *rk_c) {
+        TEST_SAY("Waiting for partition assignment\n");
+        while (1) {
+                rd_kafka_topic_partition_list_t *parts = NULL;
+
+                do_consume(rk_c, 1/*1s*/);
+
+                if (rd_kafka_assignment(rk_c->rk, &parts) !=
+                    RD_KAFKA_RESP_ERR_NO_ERROR ||
+                    !parts || parts->cnt == 0) {
+                        if (parts)
+                                rd_kafka_topic_partition_list_destroy(parts);
+                        continue;
+                }
+
+                TEST_SAY("%s got assignment of %d partition(s)\n",
+                         rd_kafka_name(rk_c->rk), parts->cnt);
+                rd_kafka_topic_partition_list_destroy(parts);
+                break;
+        }
+}
+
+static void rebalance_cb (rd_kafka_t *rk,
+                          rd_kafka_resp_err_t err,
+                          rd_kafka_topic_partition_list_t *parts,
+                          void *opaque) {
+
+        _consumer_t *rk_c = opaque;
+
+        TEST_SAY("%s rebalance #%d/%d: %s: %d partition(s)\n",
+                 rd_kafka_name(rk),
+                 rk_c->assign_cnt, rk_c->max_rebalance_cnt,
+                 rd_kafka_err2name(err),
+                 parts->cnt);
+
+        switch (err)
+        {
+        case RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS:
+                rk_c->assign_cnt++;
+
+                rd_kafka_assign(rk, parts);
+
+                TEST_ASSERT(rk_c->assign_cnt <= rk_c->max_rebalance_cnt,
+                            "%s rebalanced %d times, max was %d",
+                            rd_kafka_name(rk),
+                            rk_c->assign_cnt, rk_c->max_rebalance_cnt);
+                break;  
+
+        case RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS:
+                rd_kafka_assign(rk, NULL);
+                break;
+
+        default:
+                TEST_FAIL("rebalance failed: %s", rd_kafka_err2str(err));
+                break;
+        }
+
+}
+
+#define _CONSUMER_CNT 2
+int main_0102_static_group_rebalance (int argc, char **argv) {
+        rd_kafka_conf_t *conf;
+
+        const int msgcnt = 100;
+        const char *topic = test_mk_topic_name("0101_static_group_rebalance", 1);
+        _consumer_t c[_CONSUMER_CNT] = RD_ZERO_INIT;
+
+        test_conf_init(&conf, NULL, 60);
+
+        test_conf_set(conf, "session.timeout.ms", "1000");
+        test_conf_set(conf, "auto.offset.reset", "earliest");
+        test_conf_set(conf, "topic.metadata.refresh.interval.ms", "500");
+
+        rd_kafka_conf_set_opaque(conf, &c[0]);
+        test_conf_set(conf, "group.instance.id", "consumer1");
+        c[0].rk = test_create_consumer("static", &rebalance_cb,
```

Comment:
the group.id must be unique within the test suite since tests may run in parallel. Use the topic name.

Reply:
drop `&` from rebalance_cb